QACI-406 Websockets tck failure fix: Grizzly

dmatej opened the pull request:
- Fixes two websocket tests in TCK tests for 5.2020.5-SNAPSHOT
- Still fails one of two TrailersTests - originally failed two.
- More robust compatibility with JDK8 versions.
- build finally does not depend on https://github.com/payara/patched-src-grizzly-npn/pull/4 - grizzly npn release can be postponed.

dmatej committed:
QACI-406 Possibility to run tests with different JSSE implementations

- default is none
- profile npn uses grizzly-npn-bootstrap
- profile openjsse uses openjsse
- versions can be overriden from command line (to compare results)

dmatej committed:
QACI-406 Logging in tests

- important loggers are explicitly named in logging.properties

dmatej committed:
QACI-406 Fixed dependency artifactId

dmatej committed:
QACI-406 Fixed Tyrus dependency on HandshakeListener interface

- moving this interface out is a breaking change

dmatej committed:
QACI-406 Fixed TrailersTest - missing lock

dmatej committed:
QACI-406 Better compatibility detection

- can run with any compatible implementation, not only NPN Bootstrap
- with new JDK versions after 8u250 uses it's JSSE impl by default
- still can use older NPN bootstrap versions if configured
- can use also other implementations (openjsse)

dmatej commented on modules/grizzly/src/main/java/org/glassfish/grizzly/filterchain/DefaultFilterChain.java:
```diff
@@ -253,14 +253,14 @@ protected NextAction executeFilter(final FilterExecutor executor,
         do {
             if (LOGGER.isLoggable(Level.FINEST)) {
-                LOGGER.log(Level.FINE, "Execute filter. filter={0} context={1}",
+                LOGGER.log(Level.FINE, "before filter execution. filter={0} context={1}",
```
Symmetric logging is easier to find/grep and to understand

dmatej commented on modules/bundles/http/pom.xml:
```diff
@@ -171,7 +171,7 @@
             <groupId>org.glassfish.grizzly</groupId>
             <artifactId>grizzly-npn-api</artifactId>
-            <version>${grizzly.alpn.version}</version>
+            <version>${grizzly.npn.api.version}</version>
```
At the start I wasn't sure which property is for which dependency.

dmatej commented on modules/grizzly/src/main/java/org/glassfish/grizzly/ssl/HandshakeListener.java:
```diff
@@ -1,12 +0,0 @@
-package org.glassfish.grizzly.ssl;
```
Cause of the TCK failure. Tyrus depends on this interface, so it must not be moved outside the SSLBaseFilter class. Breaking change.

dmatej commented on modules/grizzly/src/main/java/org/glassfish/grizzly/ssl/SSLBaseFilter.java:
```diff
@@ -1208,4 +1202,15 @@ public Buffer clone(final Connection connection,
             return originalMessage;
         }
     }
+
```
Moved back.

dmatej commented on modules/http/src/main/java/org/glassfish/grizzly/http/HttpTrailer.java:
```diff
@@ -36,7 +36,7 @@
      */
     public static boolean isTrailer(HttpContent httpContent) {
-        return HttpTrailer.class.isAssignableFrom(httpContent.getClass());
+        return httpContent != null && HttpTrailer.class.isAssignableFrom(httpContent.getClass());
```
More robust. Possible NPE fixed.

dmatej commented on modules/http2/pom.xml:
```diff
@@ -128,6 +128,36 @@
-
     </dependencies>
+
+    <profiles>
```
Possible to use alternatives, but default is JDK version, not npn.
It is not even a breaking change, because the bootstrap was created to fill the gap for JDK8 versions without ALPN support and it is an **old** copy of Sun jsse classes, which were enhanced and backported back to JDK8 from JDK9/11.

dmatej commented on modules/http2/src/main/java/org/glassfish/grizzly/http2/AlpnSupport.java:
```diff
@@ -37,43 +36,24 @@
-    private static final Method nativeHandshakeMethod;
-    
-    static {
-        boolean isExtensionFound = false;
```
When I started to test combinations of JDK vs. grizzly vs. npn-bootstrap vs. npn-api, AlpnSupport sometimes threw exceptions on unexpected places. See Confluence page for few details.
Finally, I created a new class for all this heuristic and tested it against all JDK8 versions I have (both AdoptJDK and Zulu 8u265 are the newest)

dmatej commented on modules/http2/src/main/java/org/glassfish/grizzly/http2/DefaultOutputSink.java:
```diff
@@ -25,7 +25,6 @@
 import java.util.List;
```
Grizzly's tests were failing because of this class. I was able to fix one of two failing tests, but then I noticed that the lock usages are not always correct and is hard to find all paths. So I did around 100 small iterations to make the code more readable and to fix both tests.

dmatej committed:
QACI-406 Removed incorrect FIXME

dmatej commented on modules/http2/src/test/java/org/glassfish/grizzly/http2/AbstractHttp2Test.java:
```diff
@@ -45,14 +46,23 @@
  * @author Alexey Stashok
  */
 public abstract class AbstractHttp2Test {
+    static {
```
Logging in tests, so you can see order of method executions in multithreaded tests.

dmatej commented on pom.xml:
```diff
@@ -441,6 +441,8 @@
         <findbugs.common>exclude-common.xml</findbugs.common>
         <jdk.compile.version>1.8</jdk.compile.version>
-        <grizzly.alpn.version>1.9.payara-p1</grizzly.alpn.version>
+        <grizzly.npn.api.version>1.9.payara-p2</grizzly.npn.api.version>
```
Default dependency versions for actual JDK8u265

dmatej committed:
QACI-406 Released grizzly 2.4.4.payara-p4

- fixed ALPN support
- fixed two failing tests (trailers)
- better compatibility and logging

MattGill98 reviewed (changes requested):
Have you tested for regressions against PAYARA-3031?

dmatej commented:
> Have you tested for regressions against PAYARA-3031?

Nope, but I will add the reproducer to Payara Samples now. I have seen your 2 years old fix now, now I'm curious if it still works or not. But those locks before both our changes were used incorrectly.

dmatej committed:
QACI-406 Better logging, reliable locking, variable names, but ...

- in this state grizzly:
  - still fails TrailersTest (grizzly-http2)
  - passes TCK tests (websocket) (THIS CHANGED!)
  - fails h2spec (race condition)
  - fails LargeResponseTest (Payara Samples - http2, race condition)
- see lines with FIXME: this causes data corruption!

MattGill98 committed:
Remove SNAPSHOT BOM dependency

SNAPSHOT version missed in version update commit

MattGill98 commented:
I've pushed a minor update to the bom version - worth remembering in future that the `mvn versions:set` command needs running in the top level folder as well as the BOM folder

MattGill98 reviewed (approved):
Looks like a good change - make sure to push upstream when you get a chance
